src/BinSkim.Rules/SignSecurely.cs

```diff
@@ -0,0 +1,456 @@
+        /// signed with a recognized certificate (driven by a signature hash
+        /// allow list).
+        /// </summary>
+
+        public override string FullDescription
+        {
+            get { return RuleResources.BA2022_SignCorrectly_Description; }
+        }
+
+        protected override IEnumerable<string> FormatSpecifierIds
+        {
+            get
+            {
+                return new string[] {
+                    nameof(RuleResources.BA2022_Pass),
+                    nameof(RuleResources.BA2022_Fail_Sha1Signature),
+                    nameof(RuleResources.BA2022_Fail_VerifyActionFailed),
+                    nameof(RuleResources.BA2022_Fail_WinTrustVerifyApiError),
+                    nameof(RuleResources.BA2022_InvalidSignatureOrFileOpenError)};
+            }
+        }
+
+        public override AnalysisApplicability CanAnalyze(BinaryAnalyzerContext context, out string reasonForNotAnalyzing)
+        {
+            reasonForNotAnalyzing = null;
+            return AnalysisApplicability.ApplicableToSpecifiedTarget;
+        }
+
+        public override void Analyze(BinaryAnalyzerContext context)
+        {
+            Native.CERT_INFO certInfo;
+            Native.WINTRUST_DATA winTrustData;
+            string filePath = context.PE.FileName;
+
+            if (InvokeVerifyAction(context, filePath, out winTrustData) &&
+                ValidateSignatureAlgorithm(context, winTrustData, out certInfo) &&
+                VerifyCodeIsSignedByApprovedCertificate(certInfo))
+            {
+
+                // '{0}' appears to be signed securely by a trusted publisher with
+                // no verification or time stamp errors. Revocation checking was
+                // performed on the entire certificate chain, excluding the root
+                // certificate. The image was signed with '{1}', a
+                // cryptographically strong algorithm.
+                context.Logger.Log(this,
+                    RuleUtilities.BuildResult(ResultKind.Pass, context, null,
+                        nameof(RuleResources.BA2022_Pass)));
+            }
+        }
+
+        private bool InvokeVerifyAction(BinaryAnalyzerContext context, string filePath, out Native.WINTRUST_DATA winTrustData)
+        {
+            Guid action;
+            bool continueProcessing;
+            ValidationResult validationResult;
+
+            continueProcessing = true;
+            winTrustData = InitializeWinTrustDataStruct(filePath);
+
+            // First, we will invoke the basic verification. Note that currently this code path
+            // does not reach across the network to perform its function. We could optionally
+            // enable this (which would require altering the code that initializes our
+            // WINTRUST_DATA instance).
+            action = Native.ActionGenericVerifyV2;
+
+            validationResult = WinVerifyTrustHelper(Native.INVALID_HANDLE_VALUE, ref winTrustData);
```

Comment:
This won't work for things that are catalog signed, and you're leaking the WinTrust context structure. If you want to validate the signature you need to add the catalog signing check and call WinVerifyTrust with `WTD_STATEACTION_CLOSE`. You can see "Bill's Really Crappy Implementation He Wrote Some Time Ago" in C++ here: https://github.com/BillyONeal/pevFind/blob/master/pevLib/fileData.cpp#L364